mbed/mbed.py

```diff
@@ -1336,6 +1342,103 @@ def set_cache(self, url):
                 warning("Unable to cache \"%s\" to \"%s\"" % (self.path, cpath))
         return False
 
+    def cache_lock(self, url):
+        cpath = self.url2cachedir(url)
+        if not cpath:
+            return False
+
+        if not os.path.isdir(cpath):
+            os.makedirs(cpath)
+
+        lock_dir = os.path.join(cpath, '.lock')
+        lock_file = os.path.join(lock_dir, 'pid')
+        timeout = 300
+
+        for i in range(timeout): 
+            if i:
+                time.sleep(1)
+
+            if os.path.exists(lock_dir):
+                try:
+                    if os.path.isfile(lock_file):
+                        with open(lock_file, 'r', 0) as f:
+                            pid = f.read(8)
```

Comment:
Concurrency issue here.
Other proces might have just opened the file, but not yet written to it.
On line 1385 it does:
```
                    with open(lock_file, 'wb', 0) as f:
                        info("Writing cache lock file %s for pid %s" % (lock_file, os.getpid()))
                        f.write(str(os.getpid()))
                        f.flush()
                        os.fsync(f)
```
So context switch might happen after `open()` but just before writing to a file.

This would lead to two processes thinking that they own the lock.